From e850e89f3007385c2c93b82da06d10b69559e186 Mon Sep 17 00:00:00 2001
From: madroach <madroach@gmerlin.de>
Date: Wed, 13 Nov 2019 17:51:04 +0000
Subject: allow autoview preview of multipart/alternative content

add view-alt function, so now we have
 * view-alt "v"
 * view-alt-text "Esc v"
 * view-alt-mailcap "V"

Upstream-commit: https://gitlab.com/muttmua/mutt/commit/e850e89f3007385c2c93b82da06d10b69559e186
Co-authored-by: Richard Russon <rich@flatcap.org>
---
 OPS                 |  1 +
 compose.c           | 17 ++++++++++++++---
 doc/manual.xml.head |  9 +++++----
 functions.h         |  3 ++-
 4 files changed, 22 insertions(+), 8 deletions(-)

diff --git a/OPS b/OPS
index 990bdb35..46a63cf2 100644
--- a/OPS
+++ b/OPS
@@ -49,6 +49,7 @@ OP_COMPOSE_TO_SENDER "compose new message to the current message sender"
 OP_COMPOSE_TOGGLE_DISPOSITION "toggle disposition between inline/attachment"
 OP_COMPOSE_TOGGLE_UNLINK "toggle whether to delete file after sending it"
 OP_COMPOSE_UPDATE_ENCODING "update an attachment's encoding info"
+OP_COMPOSE_VIEW_ALT "view multipart/alternative"
 OP_COMPOSE_VIEW_ALT_TEXT "view multipart/alternative as text"
 OP_COMPOSE_VIEW_ALT_MAILCAP "view multipart/alternative using mailcap"
 OP_COMPOSE_WRITE_MESSAGE "write the message to a folder"
diff --git a/compose.c b/compose.c
index 6c7948e6..3aa7817c 100644
--- a/compose.c
+++ b/compose.c
@@ -1509,6 +1509,7 @@ int mutt_compose_menu(struct Email *msg,   /* structure for new message */
         mutt_message_hook(NULL, msg, MUTT_SEND2HOOK);
         break;
 
+      case OP_COMPOSE_VIEW_ALT:
       case OP_COMPOSE_VIEW_ALT_TEXT:
       case OP_COMPOSE_VIEW_ALT_MAILCAP:
       {
@@ -1522,9 +1523,19 @@ int mutt_compose_menu(struct Email *msg,   /* structure for new message */
         alternative = mutt_run_send_alternative_filter(msg->content);
         if (!alternative)
           break;
-        mutt_view_attachment(NULL, alternative,
-                              op == OP_COMPOSE_VIEW_ALT_TEXT ? MUTT_AS_TEXT : MUTT_MAILCAP,
-                              NULL, actx);
+        switch(op)
+        {
+          case OP_COMPOSE_VIEW_ALT_TEXT:
+            op = MUTT_AS_TEXT;
+            break;
+          case OP_COMPOSE_VIEW_ALT_MAILCAP:
+            op = MUTT_MAILCAP;
+            break;
+          default:
+            op = MUTT_REGULAR;
+            break;
+        }
+        mutt_view_attachment(NULL, alternative, op, NULL, actx);
         mutt_free_body(&alternative);
         break;
       }
diff --git a/doc/manual.xml.head b/doc/manual.xml.head
index 256c2a64..ed70c710 100644
--- a/doc/manual.xml.head
+++ b/doc/manual.xml.head
@@ -8157,10 +8157,11 @@ Content in html format
 </screen>
 
 <para>
-A preview of the alternative can be viewed in the compose menu
-using the functions <literal>&lt;view-alt-text&gt;</literal> and
-<literal>&lt;view-alt-mailcap&gt;</literal>, bound to &quot;v&quot;
-and &quot;V&quot; by default.
+A preview of the alternative can be viewed in the compose menu using the functions
+<literal>&lt;view-alt&gt;</literal>,
+<literal>&lt;view-alt-text&gt;</literal> and
+<literal>&lt;view-alt-mailcap&gt;</literal>, bound to
+&quot;v&quot;, &quot;Esc v&quot; and &quot;V&quot; by default.
 </para>
 
 </sect1>
diff --git a/functions.h b/functions.h
index 8f98fdb2..dc0a1c1d 100644
--- a/functions.h
+++ b/functions.h
@@ -384,7 +384,8 @@ const struct binding_t OpCompose[] = { /* map: compose */
   { "toggle-recode",    OP_COMPOSE_TOGGLE_RECODE,        NULL },
   { "update-encoding",        OP_COMPOSE_UPDATE_ENCODING,        "U" },
   { "view-attach",        OP_VIEW_ATTACH,                        MUTT_ENTER_S },
-  { "view-alt-text",    OP_COMPOSE_VIEW_ALT_TEXT,       "v" },
+  { "view-alt",                OP_COMPOSE_VIEW_ALT,                "v" },
+  { "view-alt-text",    OP_COMPOSE_VIEW_ALT_TEXT,       "\033v" },
   { "view-alt-mailcap", OP_COMPOSE_VIEW_ALT_MAILCAP,    "V" },
   { "send-message",        OP_COMPOSE_SEND_MESSAGE,        "y" },
   { "pipe-entry",        OP_PIPE,                        "|" },
